RadixWallet/Features/CreateAccount/Coordinator/CreateAccountCoordinator+Reducer.swift

```diff
@@ -112,33 +145,137 @@ extension CreateAccountCoordinator {
+			state.name = accountName
+			if useLedgerAsFactorSource {
+				state.path.append(.selectLedger(.init(context: .createHardwareAccount)))
+				return .none
+			} else {
+				return .send(.internal(.derivePublicKeyFromDevice))
+			}
+
+		case let .path(.element(_, action: .selectLedger(.delegate(.choseLedger(ledger))))):
+			return .send(.internal(.derivePublicKeyFromLedger(ledger)))
+
+		case .path(.element(_, action: .completion(.delegate(.completed)))):
+			return .run { send in
+				await send(.delegate(.completed))
+				if isPresented {
+					await dismiss()
+				}
+			}
+
+		default:
+			return .none
+		}
+	}
+
+	public func reduce(into state: inout State, internalAction: InternalAction) -> Effect<Action> {
+		switch internalAction {
+		case .derivePublicKeyFromDevice:
+			state.destination = .derivePublicKey(
+				.init(
+					derivationPathOption: .next(
+						for: .account,
+						networkID: state.config.specificNetworkID,
+						curve: .curve25519,
+						scheme: .cap26
+					),
+					factorSourceOption: .device,
+					purpose: .createNewEntity(kind: .account)
+				))
+			return .none
+
+		case let .derivePublicKeyFromLedger(ledger):
+			state.destination = .derivePublicKey(
+				.init(
+					derivationPathOption: .next(
+						for: .account,
+						networkID: state.config.specificNetworkID,
+						curve: .curve25519,
+						scheme: .cap26
+					),
+					factorSourceOption: .specific(ledger.embed()),
+					purpose: .createNewEntity(kind: .account)
+				))
 			return .none
 
-		case let .path(.element(_, action: .step2_creationOfAccount(.delegate(.createdAccount(newAccount))))):
-			state.path.append(.step3_completion(.init(
-				account: newAccount,
+		case let .createAccountResult(.success(account)):
+			return .run { send in
+				try await accountsClient.saveVirtualAccount(account)
+				await send(.internal(.handleAccountCreated(account)))
+			} catch: { error, send in
+				loggerGlobal.error("Failed to save newly created virtual account into profile: \(error)")
+				await send(.internal(.handleFailure))
+			}
+
+		case let .createAccountResult(.failure(error)):
+			errorQueue.schedule(error)
+			state.destination = nil
+			return .none
+
+		case let .handleAccountCreated(account):
+			state.destination = nil
+			state.path.append(.completion(.init(
+				account: account,
 				config: state.config
 			)))
 			return .none
 
-		case .path(.element(_, action: .step2_creationOfAccount(.delegate(.createAccountFailed)))):
-			state.path.removeLast()
+		case .handleFailure:
+			state.destination = nil
 			return .none
+		}
+	}
+
+	public func reduce(into state: inout State, presentedAction: Destination.Action) -> Effect<Action> {
+		switch presentedAction {
+		case let .derivePublicKey(.delegate(.derivedPublicKeys(hdKeys, factorSourceID, networkID))):
+			guard let hdKey = hdKeys.first else {
+				loggerGlobal.error("Failed to create account expected one single key, got: \(hdKeys.count)")
+				return .send(.internal(.handleFailure))
```

Comment:
In general, this seems wasteful, a function should be called directly which does the necessary job, the `internal` actions should be emitted from an Effect.